k3k-kubelet/provider/provider.go

```diff
@@ -193,75 +221,223 @@ func (p *Provider) PortForward(ctx context.Context, namespace, pod string, port
 // state, as well as the pod. DeletePod may be called multiple times for the same pod.
 func (p *Provider) DeletePod(ctx context.Context, pod *corev1.Pod) error {
-	err := p.CoreClient.Pods(p.ClusterNamespace).Delete(ctx, p.translateTo(pod).Name, metav1.DeleteOptions{})
+	p.logger.Infof("Got request to delete pod %s", pod.Name)
+	hostName := p.Translater.TranslateName(pod.Namespace, pod.Name)
+	err := p.CoreClient.Pods(p.ClusterNamespace).Delete(ctx, hostName, metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to delete pod %s/%s: %w", pod.Namespace, pod.Name, err)
+	}
+	err = p.pruneUnusedVolumes(ctx, pod)
+	if err != nil {
+		// note that we don't return an error here. The pod was sucessfully deleted, another process
+		// should clean this without affecting the user
+		p.logger.Errorf("failed to prune leftover volumes for %s/%s: %w, resources may be left", pod.Namespace, pod.Name, err)
+	}
 	p.logger.Infof("Deleted pod %s", pod.Name)
-	return err
+	return nil
+}
+
+// pruneUnusedVolumes removes volumes in use by pod that aren't used by any other pods
+func (p *Provider) pruneUnusedVolumes(ctx context.Context, pod *corev1.Pod) error {
+	rawSecrets, rawConfigMaps := getSecretsAndConfigmaps(pod)
+	// since this pod was removed, originally mark all of the secrets/configmaps it uses as eligible
+	// for pruning
+	pruneSecrets := sets.Set[string]{}.Insert(rawSecrets...)
+	pruneConfigMap := sets.Set[string]{}.Insert(rawConfigMaps...)
+	var pods corev1.PodList
+	// only pods in the same namespace could be using secrets/configmaps that this pod is using
+	err := p.VirtualClient.List(ctx, &pods, &client.ListOptions{
+		Namespace: pod.Namespace,
+	})
+	if err != nil {
+		return fmt.Errorf("unable to list pods: %w", err)
+	}
+	for _, vPod := range pods.Items {
+		if vPod.Name == pod.Name {
+			continue
+		}
+		secrets, configMaps := getSecretsAndConfigmaps(&vPod)
+		pruneSecrets.Delete(secrets...)
+		pruneConfigMap.Delete(configMaps...)
+	}
+	for _, secretName := range pruneSecrets.UnsortedList() {
+		var secret corev1.Secret
+		err := p.VirtualClient.Get(ctx, types.NamespacedName{
+			Name:      secretName,
+			Namespace: pod.Namespace,
+		}, &secret)
+		if err != nil {
+			return fmt.Errorf("unable to get secret %s/%s for pod volume: %w", pod.Namespace, secretName, err)
+		}
+		err = p.Handler.RemoveResource(ctx, &secret)
+		if err != nil {
+			return fmt.Errorf("unable to remove secret %s/%s for pod volume: %w", pod.Namespace, secretName, err)
+		}
+	}
+	for _, configMapName := range pruneConfigMap.UnsortedList() {
+		var configMap corev1.ConfigMap
+		err := p.VirtualClient.Get(ctx, types.NamespacedName{
+			Name:      configMapName,
+			Namespace: pod.Namespace,
+		}, &configMap)
+		if err != nil {
+			return fmt.Errorf("unable to get configMap %s/%s for pod volume: %w", pod.Namespace, configMapName, err)
+		}
+		err = p.Handler.RemoveResource(ctx, &configMap)
```

Comment:
since this is not really a controller the k8s api can return an error if the resource is already locked by any other operation and virtual kubelet may create an error then, I think we can either add a retry-backoff mechanism or use a thread to delete the pod?

Reply:
Retry would be cleaner and likely more inline with k8s code expectations.

Reply:
I'm ok with adding that elsewhere, but I don't think we can add that here. The provider doesn't "know" how to create/update/or delete resources (other than pods) in the host cluster - the controllers and the handlers are what know how to do that.